File: JavaPrototypes/ExpenseCalculator/tests/com/example/expenseCalculatorTests/DepartmentTests.java
package com.example.expenseCalculatorTests;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import com.example.expenseCalculator.Department;
import com.example.expenseCalculator.Employee;
import com.example.expenseCalculator.EmployeeFactory;
import com.example.expenseCalculator.EmployeeType;
import com.example.expenseCalculator.Manager;

public class DepartmentTests {

	@Before
	public void setUp() throws Exception {
	}

	@After
	public void tearDown() throws Exception {
	}
	
	private static ExpenseCalculatorTestHelper testHelper = new ExpenseCalculatorTestHelper();
	private Pair<String, Double> _qatester = testHelper.findPairByString("QA Tester");
	private Pair<String, Double> _developer = testHelper.findPairByString("Developer");
	private Pair<String, Double> _manager = testHelper.findPairByString("Manager");

	@Test
	public void DepartmentTotalExpenses_Equals_TotalEmployeeExpenses() {
		
		Department marketing = new Department() {{
				setName("Markteting");
			}};
		
		Employee employeeQA = EmployeeFactory.create(EmployeeType.Employee, _qatester.x, _qatester.y);
		Employee employeeDEV = EmployeeFactory.create(EmployeeType.Employee, _developer.x, _developer.y);
		Manager employeeMGR = (Manager) EmployeeFactory.create(EmployeeType.Manager, _manager.x, _manager.y);
	
		employeeMGR.getEmployees().add(employeeQA);
		employeeMGR.getEmployees().add(employeeDEV);
		
		marketing.getEmployees().add(employeeMGR);
		
		Assert.assertEquals(testHelper.getTotalExpenses(), marketing.getTotalExpenses(), 0);
	}
	
	@Test
	public void DepartmentWithStandAloneEmployeesTotalExpenses_Equals_TotalEmployeeExpenses() {
		
		Department marketing = new Department() {{
				setName("Markteting");
			}};
		
		Employee employeeQA = EmployeeFactory.create(EmployeeType.Employee, _qatester.x, _qatester.y);
		Employee employeeDEV = EmployeeFactory.create(EmployeeType.Employee, _developer.x, _developer.y);
		Manager employeeMGR = (Manager) EmployeeFactory.create(EmployeeType.Manager, _manager.x, _manager.y);
	
		employeeMGR.getEmployees().add(employeeQA);
		employeeMGR.getEmployees().add(employeeDEV);
		
		double manager1Expenses = testHelper.getTotalExpenses();
		
		Employee employeeQA1 = EmployeeFactory.create(EmployeeType.Employee, _qatester.x, _qatester.y);
		Employee employeeDEV1 = EmployeeFactory.create(EmployeeType.Employee, _developer.x, _developer.y);
		
		marketing.getEmployees().add(employeeQA1);
		marketing.getEmployees().add(employeeDEV1);
		marketing.getEmployees().add(employeeMGR);
		
		Assert.assertEquals(manager1Expenses + employeeQA1.getExpenseAllocation() + employeeDEV1.getExpenseAllocation(), marketing.getTotalExpenses(),  0);
	}
	
	@Test
	public void DepartmentWithNestedManagerTotalExpenses_Equals_TotalEmployeeExpenses() {
		
		Department marketing = new Department() {{
			setName("Markteting");
		}};
		
		Employee employeeQA = EmployeeFactory.create(EmployeeType.Employee, _qatester.x, _qatester.y);
		Employee employeeDEV = EmployeeFactory.create(EmployeeType.Employee, _developer.x, _developer.y);
		Manager employeeMGR = (Manager) EmployeeFactory.create(EmployeeType.Manager, _manager.x, _manager.y);
	
		employeeMGR.getEmployees().add(employeeQA);
		employeeMGR.getEmployees().add(employeeDEV);
		
		double manager1Expenses = testHelper.getTotalExpenses();
		
		Employee employeeQA1 = EmployeeFactory.create(EmployeeType.Employee, _qatester.x, _qatester.y);
		Employee employeeDEV1 = EmployeeFactory.create(EmployeeType.Employee, _developer.x, _developer.y);
		Manager employeeMGR1 = (Manager) EmployeeFactory.create(EmployeeType.Manager, _manager.x, _manager.y);
		
		employeeMGR1.getEmployees().add(employeeQA1);
		employeeMGR1.getEmployees().add(employeeDEV1);
		
		double manager2Expenses = testHelper.getTotalExpenses();
		
		employeeMGR1.getEmployees().add(employeeMGR);
		
		
		double totalExpenses = manager1Expenses + manager2Expenses;
		
		marketing.getEmployees().add(employeeMGR1);
		
		Assert.assertEquals(totalExpenses, marketing.getTotalExpenses(), 0);
	}

}

File: DotNetPrototypes/ExpenseCalculatorTests/DepartmentTests.cs
using System;
using ExpenseCalculator;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace ExpenseCalculatorTests
{
    [TestClass]
    public class DepartmentTests
    {
        private static ExpenseCalculatorTestHelper testHelper = new ExpenseCalculatorTestHelper();
        private Tuple<string, decimal> _qatester = testHelper.Employees.Find(delegate(Tuple<string, decimal> t) { return t.Item1 == "QA Tester"; });
        private Tuple<string, decimal> _developer = testHelper.Employees.Find(delegate(Tuple<string, decimal> t) { return t.Item1 == "Developer"; });
        private Tuple<string, decimal> _manager = testHelper.Employees.Find(delegate(Tuple<string, decimal> t) { return t.Item1 == "Manager"; });

        [TestMethod]
        public void DepartmentTotalExpenses_Equals_TotalEmployeeExpenses()
        {
            Department marketing = new Department { 
                Name = "Markteting"
            };

            IEmployee qaTester = EmployeeFactory.Create(EmployeeType.Employee, _qatester.Item1, _qatester.Item2);
            IEmployee developer = EmployeeFactory.Create(EmployeeType.Employee, _developer.Item1, _developer.Item2);
            Manager manager = (Manager)EmployeeFactory.Create(EmployeeType.Manager, _manager.Item1, _manager.Item2);

            manager.Employees.Add(qaTester);
            manager.Employees.Add(developer);

            marketing.Employees.Add(manager);

            Assert.AreEqual(marketing.GetTotalExpenses(), testHelper.GetTotalExpenses());
        }

        [TestMethod]
        public void DepartmentWithStandAloneEmployeesTotalExpenses_Equals_TotalEmployeeExpenses()
        {
            Department marketing = new Department
            {
                Name = "Markteting"
            };

            IEmployee qaTester1 = EmployeeFactory.Create(EmployeeType.Employee, _qatester.Item1, _qatester.Item2);
            IEmployee developer1 = EmployeeFactory.Create(EmployeeType.Employee, _developer.Item1, _developer.Item2);
            Manager manager = (Manager)EmployeeFactory.Create(EmployeeType.Manager, _manager.Item1, _manager.Item2);

            manager.Employees.Add(qaTester1);
            manager.Employees.Add(developer1);

            decimal manager1Expenses = testHelper.GetTotalExpenses();

            IEmployee qaTester2 = EmployeeFactory.Create(EmployeeType.Employee, _qatester.Item1, _qatester.Item2);
            IEmployee developer2= EmployeeFactory.Create(EmployeeType.Employee, _developer.Item1, _developer.Item2);

            marketing.Employees.Add(manager);
            marketing.Employees.Add(qaTester2);
            marketing.Employees.Add(developer2);

            Assert.AreEqual(marketing.GetTotalExpenses(), manager1Expenses + qaTester2.ExpenseAllocation + developer2.ExpenseAllocation);
        }

        [TestMethod]
        public void DepartmentWithNestedManagerTotalExpenses_Equals_TotalEmployeeExpenses()
        {
            Department marketing = new Department
            {
                Name = "Markteting"
            };

            IEmployee qaTester1 = EmployeeFactory.Create(EmployeeType.Employee, _qatester.Item1, _qatester.Item2);
            IEmployee developer1 = EmployeeFactory.Create(EmployeeType.Employee, _developer.Item1, _developer.Item2);
            Manager manager1 = (Manager)EmployeeFactory.Create(EmployeeType.Manager, _manager.Item1, _manager.Item2);

            manager1.Employees.Add(qaTester1);
            manager1.Employees.Add(developer1);

            decimal manager1Expenses = testHelper.GetTotalExpenses();

            IEmployee qaTester2 = EmployeeFactory.Create(EmployeeType.Employee, _qatester.Item1, _qatester.Item2);
            IEmployee developer2 = EmployeeFactory.Create(EmployeeType.Employee, _developer.Item1, _developer.Item2);
            Manager manager2 = (Manager)EmployeeFactory.Create(EmployeeType.Manager, _manager.Item1, _manager.Item2);

            manager2.Employees.Add(qaTester2);
            manager2.Employees.Add(developer2);

            decimal manager2Expenses = testHelper.GetTotalExpenses();

            manager2.Employees.Add(manager1);

            marketing.Employees.Add(manager2);

            Assert.AreEqual(marketing.GetTotalExpenses(), manager1Expenses + manager2Expenses);
        }

    }
}
